### PR TITLE
feat: add ONNX inference backend with build tag gating

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

- Adds ONNX Runtime inference support as an optional backend, gated behind the `onnx` build tag
- Vendors `go-birdnet-onnx` library as `internal/inference/onnx/` sub-package supporting BirdNET v2.4, v3.0, and Google Perch v2 models
- Auto-selects backend based on model file extension (`.onnx` → ONNX, otherwise TFLite)
- Default builds have zero ONNX dependency — builds without `-tags onnx` return a clear error for `.onnx` model paths
- Adds `PredictRaw()` methods to return pre-activation logits so BirdNET-Go's sensitivity-adjusted sigmoid handles post-processing
- Adds `ONNXRuntimePath` config field for specifying the ONNX Runtime shared library path

Builds on #2617 which established the `inference.Classifier` and `inference.RangeFilter` interfaces.

## Test Plan

- [x] `go build -tags noembed,skipfrontend ./...` — default build (no ONNX) clean
- [x] `go build -tags onnx,noembed,skipfrontend ./...` — ONNX build clean
- [x] `go test -race -tags noembed,skipfrontend ./internal/birdnet/...` — all tests pass
- [x] `golangci-lint run --build-tags=onnx,noembed,skipfrontend` — zero issues
- [x] `go-tflite` imports isolated to `internal/inference/tflite.go`
- [x] `onnxruntime_go` imports isolated to `internal/inference/onnx/` (build-tag gated)

## Summary by CodeRabbit

* **New Features**
  * ONNX model support as an alternative inference backend with automatic model-format detection.
  * New configuration option to specify an ONNX runtime library path.
  * Range-filter models supported for all backends.
  * Batched inference and richer prediction outputs (embeddings, top‑K, confidence).

* **Improvements**
  * Robust label-file loading for .txt/.csv/.json with clearer validation and error messages.
  * Improved input validation and more descriptive inference errors.

### CODE REVIEW

File: internal/inference/onnx/classifier.go [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+		return nil, fmt.Errorf("birdnet: failed to create session options: %w", err)
+	}
+	defer func() { _ = sessOpts.Destroy() }()
+
+	if err := sessOpts.SetIntraOpNumThreads(1); err != nil {
+		return nil, fmt.Errorf("birdnet: failed to set intra-op threads: %w", err)
+	}
+	if err := sessOpts.SetInterOpNumThreads(1); err != nil {
+		return nil, fmt.Errorf("birdnet: failed to set inter-op threads: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx/classifier.go:201]: Fixed in e2a4eab5 — thread count from config now wired through to ONNX session options via WithSessionOptions callback.

File: internal/inference/onnx.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+// Must be called before creating any ONNX classifiers or range filters.
+func InitONNXRuntime(libraryPath string) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("failed to initialize ONNX Runtime: %v", r)
+		}
+	}()
+	ort.MustInitORT(libraryPath)
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx.go]: The ONNX Runtime itself is a process-global singleton by design (ort.InitializeEnvironment/DestroyEnvironment). Wrapping it in a struct doesn't change this — it's still global state underneath. The current approach is explicit about this constraint. The panic recovery in InitONNXRuntime already handles the failure case gracefully.

File: internal/birdnet/birdnet.go [hunk 3]
```diff
@@ -154,8 +154,24 @@ func NewBirdNET(settings *conf.Settings) (*BirdNET, error) {
 	return bn, nil
 }
 
+// isONNXModel returns true if the model path points to an ONNX model file.
+func isONNXModel(path string) bool {
+	return strings.HasSuffix(strings.ToLower(path), ".onnx")
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/birdnet/birdnet.go:164]: Fixed in e2a4eab5 — isONNXModel() now expands environment variables before checking the extension.

File: internal/birdnet/birdnet.go [hunk 4]
```diff
@@ -154,8 +154,24 @@ func NewBirdNET(settings *conf.Settings) (*BirdNET, error) {
...
+
 // initializeModel loads and initializes the primary BirdNET model.
+// Dispatches to ONNX or TFLite backend based on the model file extension.
 func (bn *BirdNET) initializeModel() error {
+	// If model path ends with .onnx, use the ONNX backend
+	if isONNXModel(bn.Settings.BirdNET.ModelPath) {
+		return bn.initializeONNXModel()
+	}
+
+	return bn.initializeTFLiteModel()
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/birdnet/birdnet.go:174]: Good catch — fixed in e2a4eab5 by moving loadLabels() before initializeModel() in NewBirdNET(). ONNX models need labels at construction time for output dimension validation, unlike TFLite which validates later.

File: internal/birdnet/model_onnx.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+			Category(errors.CategoryModelInit).
+			Context("model_type", "range_filter").
+			Context("range_filter_model", bn.Settings.BirdNET.RangeFilter.ModelPath).
+			Timing("onnx-meta-model-init", time.Since(start)).
+			Build()
+	}
+
+	bn.rangeFilter = rangeFilter
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/birdnet/model_onnx.go:78]: Addressed by the same fix — loadLabels() now runs before both initializeModel() and initializeMetaModel(), so labels are available for ONNX meta model init.

File: internal/inference/onnx/classifier.go [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+	logitsDims := outputInfos[modelCfg.LogitsIndex].Dimensions
+	if len(logitsDims) < 2 {
+		return nil
+	}
+	logitsSize := int(logitsDims[len(logitsDims)-1])
+	if logitsSize > 0 && labelCount != logitsSize {
+		return &LabelCountError{Expected: logitsSize, Got: labelCount}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx/classifier.go:185]: Fixed in e2a4eab5 — validateLabelCount now returns an error when LogitsIndex is out of range instead of silently skipping validation.

File: internal/inference/onnx/classifier.go [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+	if cfg.labelsPath != "" {
+		return loadLabels(cfg.labelsPath)
+	}
+	return nil, ErrLabelsRequired
+}
+
+// Config returns the model configuration.
+func (c *Classifier) Config() ModelConfig {
+	return c.config
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx/classifier.go:230]: Fixed in e2a4eab5 — Config() now deep copies InputShape to prevent aliasing.

File: internal/inference/onnx/labels.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+		}
+	}
+	for i, h := range header {
+		h = strings.TrimSpace(h)
+		if _, err := strconv.Atoi(h); err != nil {
+			return i
+		}
+	}
+	return 0
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx/labels.go:102]: Fixed in e2a4eab5 — findLabelColumn now returns an error when no recognized column header is found instead of guessing the first non-numeric column.

File: internal/inference/onnx/rangefilter.go [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+	}
+
+	return &RangeFilter{
+		session:    session,
+		labels:     labels,
+		threshold:  cfg.threshold,
+		inputName:  inputNames[0],
+		outputName: outputNames[0],
+	}, nil
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx/rangefilter.go:94]: Fixed in e2a4eab5 — NewRangeFilter now validates label count against the model output dimension and returns LabelCountError on mismatch.

File: internal/birdnet/birdnet.go [hunk 10]
```diff
@@ -154,8 +156,26 @@ func NewBirdNET(settings *conf.Settings) (*BirdNET, error) {
 	return bn, nil
 }
 
+// isONNXModel returns true if the model path points to an ONNX model file.
+// Expands environment variables before checking the extension.
+func isONNXModel(path string) bool {
+	expanded := os.ExpandEnv(path)
+	return strings.HasSuffix(strings.ToLower(expanded), ".onnx")
+}
+
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/birdnet/birdnet.go:165]: Fixed in 7c1d0540 — loadLabels() now runs before initializeModel() in ReloadModel(), matching the order established in NewBirdNET(). Same root cause as the initial-load fix.

[the remark this answers is not in the document]
Comment (tphakala) [reply internal/birdnet/birdnet.go:165]: Already addressed — loadLabels() was moved before initializeModel() in both NewBirdNET() and ReloadModel() in the first fix push.

File: internal/birdnet/model_onnx.go [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+		logger.Int("species", classifier.NumSpecies()))
+
+	return nil
+}
+
+// initializeONNXMetaModel loads and initializes an ONNX range filter meta model.
+func (bn *BirdNET) initializeONNXMetaModel() error {
+	start := time.Now()
+
+	rangeFilter, err := inference.NewONNXRangeFilter(
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/birdnet/model_onnx.go:61]: Fixed in 8486d322 — initializeONNXMetaModel now calls InitONNXRuntime before creating the range filter session.

File: internal/inference/onnx.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx.go]: Fixed in 8486d322 — InitONNXRuntime now uses sync.Once for idempotent initialization.

File: internal/inference/onnx.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+	classifierOpts := []ort.ClassifierOption{
+		ort.WithLabels(opts.Labels),
+		ort.WithTopK(0),          // We handle topK in BirdNET-Go's post-processing
+		ort.WithMinConfidence(0), // No filtering, return all raw scores
+	}
+	if opts.Threads > 0 {
+		threads := opts.Threads
+		classifierOpts = append(classifierOpts, ort.WithSessionOptions(func(so *ortlib.SessionOptions) {
+			_ = so.SetIntraOpNumThreads(threads)
+			_ = so.SetInterOpNumThreads(threads)
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx.go]: These are inside a callback passed to WithSessionOptions — the errors can't be propagated through the callback return type. The `_ =` is intentional to silence errcheck. In practice, setting thread counts never fails on valid SessionOptions.

File: internal/inference/onnx.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+// NumSpecies returns the number of species in the model output.
+func (c *onnxClassifier) NumSpecies() int {
+	return c.numSpecies
+}
+
+// Close releases the ONNX session resources.
+func (c *onnxClassifier) Close() {
+	if c.classifier != nil {
+		_ = c.classifier.Close()
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx.go:84]: The inference.Classifier interface's Close() method intentionally returns nothing (matching the TFLite backend). The `_ = c.classifier.Close()` discards the concrete ONNX type's error return. Adding error returns to Close() is a Phase 3 consideration.
